src/pyodbc.pyi

```diff
@@ -324,97 +334,587 @@ class ProgrammingError(DatabaseError): ...
+    create a Connection object.
+    """
+
+    @property
+    def autocommit(self) -> bool:
+        """Whether the database automatically executes a commit after every successful transaction.
+        Default is False.
+        """
+        ...
+
+    @autocommit.setter
+    def autocommit(self, value: bool) -> None:
+        ...
+
+    @property
+    def maxwrite(self) -> int:
+        """The maximum bytes to write before using SQLPutData, default is zero for no maximum."""
+        ...
+
+    @maxwrite.setter
+    def maxwrite(self, value: int) -> None:
+        ...
+
+    @property
+    def searchescape(self) -> str:
+        """The character for escaping search pattern characters like "%" and "_".
+        This is typically the backslash character but can be driver-specific."""
+        ...
+
+    @property
+    def timeout(self) -> int:
+        """The timeout in seconds for SQL queries, use zero (the default) for no timeout limit."""
+        ...
+
+    @timeout.setter
+    def timeout(self, value: int) -> None:
+        ...
 
-    # read-write attributes
-    autocommit: bool
-    maxwrite: int
-    timeout: int
-
-    # read-only attributes
-    searchescape: str
 
     # implemented dunder methods
     def __enter__(self) -> Connection: ...
     def __exit__(self, exc_type, exc_value, traceback) -> None: ...
 
-    # define text encoding for data, metadata, etc.
-    def setencoding(self, encoding: str, ctype: Optional[int]) -> None: ...
-    def setdecoding(self, encoding: str, ctype: Optional[int]) -> None: ...
 
-    # connection attributes
-    def getinfo(self, infotype: int, /) -> Any: ...
-    def set_attr(self, attr_id: int, value: int, /) -> None: ...
+    # functions for defining the text encoding used for data, metadata, sql, parameters, etc.
 
-    # handle non-standard database data types
-    def add_output_converter(self, sqltype: int, new_converter: Callable, /) -> None: ...
-    def get_output_converter(self, sqltype: int, /) -> Optional[Callable]: ...
-    def remove_output_converter(self, sqltype: int, /) -> None: ...
-    def clear_output_converters(self) -> None: ...
+    def setencoding(self,
+                    encoding: Optional[str] = None,
+                    ctype: Optional[int] = None) -> None:
+        """Set the text encoding for SQL statements and textual parameters sent to the database.
 
-    # query functions (in rough order of use)
-    def cursor(self) -> Cursor: ...
-    def execute(self, sql: str, *params: Any) -> Cursor: ...
-    def commit(self) -> None: ...
-    def rollback(self) -> None: ...
-    def close(self) -> None: ...
+        Args:
+            encoding: Text encoding codec, e.g. "utf-8".
+            ctype: The C data type when passing data - either pyodbc.SQL_CHAR or pyodbc.SQL_WCHAR.  More
+                relevant for Python 2.7.
+        """
+        ...
 
+    def setdecoding(self,
+                    sqltype: int,
```

Comment:
Was the Literal type annotation considered for "enum-like" values such as this one? (See [typing.Literal](https://docs.python.org/3/library/typing.html#typing.Literal))

Reply:
The Literal annotation requires literal integer values like 25, -123, etc. rather than variable names like "SQL_WCHAR".  The ODBC attributes like SQL_WCHAR and SQL_DRIVER_NAME are not defined by pyodbc as such, the values are part of the underlying ODBC protocol, and these values are generally not exposed.  For example, SQL_CHAR is defined in `sql.h` as value 1.  SQL_WCHAR is defined in `sqlucode.h` as value -8.  But nobody ever uses 1 and -8 in code.  Although defining these Python parameters as Literals is a noble aim, I'm not sure it's truly practical.

Reply:
Excellent reason, thanks a lot for your work on this project :)